paparazzi/src/main/java/app/cash/paparazzi/internal/apng/ApngReader.kt

```diff
@@ -0,0 +1,265 @@
+    buffer.close()
+  }
+
+  private fun processTo(header: Header) {
+    do {
+      val (currentHeader, data) = buffer.readChunk()
+      when (currentHeader) {
+        IHDR -> data.parseIHDR()
+        ACTL -> data.parseACTL()
+        FCTL -> data.parseFCTL()
+        IDAT -> data.parseIDAT()
+        FDAT -> data.parseFDAT()
+        IEND -> throw IllegalStateException("PNG ended while processing to $header")
+      }
+    } while (!header.bytes.contentEquals(currentHeader.bytes) && !finished())
+  }
+
+  private fun BufferedSource.parseIHDR() {
+    width = readInt()
+    height = readInt()
+    scratchSpace = IntArray(width * height)
+
+    skip(1L) // Bit depth
+    hasAlpha = when (readByte()) {
+      PNG_COLOR_TYPE_RGB -> false
+      PNG_COLOR_TYPE_RGBA -> true
+      else -> throw IllegalStateException("Unsupported color model")
+    }
+  }
+
+  private fun BufferedSource.parseACTL() {
+    frameCount = readInt()
+  }
+
+  private fun BufferedSource.parseFCTL() {
+    val nextSequence = readInt()
+    frameInfo = FrameInfo(
+      width = readInt(),
+      height = readInt(),
+      offsetX = readInt(),
+      offsetY = readInt(),
+      delayNumerator = readShort(),
+      delayDenominator = readShort()
+    )
+
+    if (sequenceNumber + 1 != nextSequence) {
+      throw IllegalStateException("Out of order sequence, expected: ${sequenceNumber + 1} actual: $nextSequence")
+    }
+    sequenceNumber = nextSequence
+  }
+
+  private fun BufferedSource.parseIDAT() {
+    decompress().decodeToScratchSpace()
+  }
+
+  private fun BufferedSource.parseFDAT() {
+    val nextSequence = readInt()
+    if (sequenceNumber + 1 != nextSequence) {
+      throw IllegalStateException("Out of order sequence, expected: ${sequenceNumber + 1} actual: $nextSequence")
+    }
+    sequenceNumber = nextSequence
+    decompress().decodeToScratchSpace()
+  }
+
+  private fun BufferedSource.decompress(): Buffer {
+    val inflater = Inflater().apply {
+      setInput(readByteArray())
+    }
+
+    val buffer = ByteArray(BUFFER_SIZE)
+    return Buffer().apply {
+      do {
+        val byteCount = inflater.inflate(buffer)
+        write(buffer, 0, byteCount)
+      } while (!inflater.finished() && byteCount != 0)
+    }
+  }
+
+  private fun BufferedSource.decodeToScratchSpace() {
+    val frameWidth = frameInfo?.width ?: width
+    val frameHeight = frameInfo?.height ?: height
+    val frameOffsetX = frameInfo?.offsetX ?: 0
+    val frameOffsetY = frameInfo?.offsetY ?: 0
+
+    for (y in 0 until frameHeight) {
+      val filterByte = readByte()
+      if (filterByte != 0.toByte()) {
+        throw IllegalStateException("Only filter None is supported, $filterByte")
+      }
+
+      for (x in 0 until frameWidth) {
+        val pixelBytes = if (hasAlpha) readByteArray(4) else readByteArray(3)
+        val rgba = Buffer().run {
+          write(pixelBytes)
+          if (!hasAlpha) {
+            writeByte(0)
+          }
+
+          readInt()
+        }
```

Comment:
Kill these allocations!
```suggestion
        val rgba = if (hasAlpha) {
          readInt()
        } else {
          (readByte() and 0xff shl 24)
            or (readByte() and 0xff shl 16)
            or (readByte() and 0xff shl 8)
        }
```